File: frontend/src/lib/api/visibility.ts
// Client-side mirror of the backend's per-field visibility chain
// (internal/visibility/visibility.go, mi-37w / mi-fo8). Pure
// function, no DOM or Svelte deps — exists so edit screens can
// surface "Inherited: owner-only (from your account default)"
// style affordances by introspecting which layer of the chain
// supplied the effective value.
//
// This is UI-affordance only. Security enforcement happens on
// the server; by the time a response reaches the SPA, redacted
// fields have already been omitted. If this file ever drifts
// from the Go helper, the table-driven tests in visibility.test.ts
// (mirrored from internal/visibility/visibility_test.go) fail and
// surface the divergence loudly. Keep the two in lock-step.

import type { components } from './schema';

export type Visibility = 'private' | 'unlisted' | 'public';

// Field names match the keys used in users.field_defaults JSON
// (see CONTRACT.md §13 v2 / mi-fo8). They are the canonical
// identifiers the chain reads from FieldDefaultsView.
export type Field = 'price' | 'acquired_from' | 'images';

// Source labels the chain layer that supplied the effective
// value. Mirrors visibility.Source in the Go helper; edit
// screens use it to phrase the affordance ('from your account
// default', 'from this specimen', etc.).
export type Source =
  | 'image'
  | 'specimen-field'
  | 'specimen-overall'
  | 'user-default'
  | 'system-default';

export interface Resolution {
  visibility: Visibility;
  source: Source;
}

// SystemDefault is the conservative fallback when no chain layer
// supplied a value. Matches visibility.SystemDefault on the Go
// side (domain.VisibilityPrivate); this is the V1→V2 default that
// keeps migrated rows from leaking on the wire post-V2.
export const SystemDefault: Visibility = 'private';

// SpecimenLike is the subset of SpecimenView the chain consumes.
// The per-field override columns (visibility_price,
// visibility_acquired_from, visibility_images) are not yet
// surfaced on SpecimenView in schema.d.ts — the edit-screen
// beads (#6/#7/#8) will add them. Declaring the input as a
// minimal interface here lets this helper compile today and
// stay stable when those fields land.
export interface SpecimenLike {
  visibility?: Visibility | null;
  visibility_price?: Visibility | null;
  visibility_acquired_from?: Visibility | null;
  visibility_images?: Visibility | null;
}

// PhotoLike is the subset of PhotoView the chain consumes.
// PhotoView.visibility is not yet surfaced in schema.d.ts for
// the same reason as SpecimenLike's override columns.
export interface PhotoLike {
  visibility?: Visibility | null;
}

// OwnerLike is the subset of ProfileBody the chain consumes —
// only the field_defaults map. Backend uses the owning user
// (the specimen's author), not the viewer; the SPA must pass
// the same.
export interface OwnerLike {
  field_defaults?: components['schemas']['FieldDefaultsView'] | null;
}

// resolveScalar resolves the visibility of a scalar redactable
// field — 'price' or 'acquired_from' — by walking the chain:
//
//   1. the specimen's per-field override
//   2. the owner's per-field default (field_defaults)
//   3. SystemDefault
//
// Passing 'images' here is a programmer error — image
// resolution requires the photo-level override and the
// specimen overall column that resolveScalar does not touch.
// The call is handled defensively (no scalar override exists
// for 'images', so the chain falls through to the owner
// default and system default) but callers SHOULD use
// resolveImage for image fields, matching the Go helper.
export function resolveScalar(field: Field, specimen: SpecimenLike, owner: OwnerLike): Resolution {
  const override = specimenScalarOverride(field, specimen);
  if (override != null) {
    return { visibility: override, source: 'specimen-field' };
  }
  const def = ownerDefault(field, owner);
  if (def != null) {
    return { visibility: def, source: 'user-default' };
  }
  return { visibility: SystemDefault, source: 'system-default' };
}

// resolveImage resolves the visibility of a single photo by
// walking the image-specific chain:
//
//   1. the photo's own visibility override
//   2. the specimen's visibility_images override
//   3. the specimen's overall visibility column
//   4. the owner's 'images' default
//   5. SystemDefault
//
// Per the mi-fo8 open question, an image override that LOOSENS
// against the specimen overall (e.g. public photo on a private
// specimen) is allowed by design. If that policy reverses,
// change happens here AND in the Go helper.
export function resolveImage(
  specimen: SpecimenLike,
  owner: OwnerLike,
  image: PhotoLike,
): Resolution {
  if (image.visibility != null) {
    return { visibility: image.visibility, source: 'image' };
  }
  if (specimen.visibility_images != null) {
    return { visibility: specimen.visibility_images, source: 'specimen-field' };
  }
  // The Go helper treats the empty string as 'not set' to cover
  // the new-specimen / not-yet-persisted construction path; the
  // SPA equivalent treats null/undefined the same way. A blank
  // specimen.visibility (zero value on the wire) falls through.
  if (specimen.visibility != null && specimen.visibility !== ('' as Visibility)) {
    return { visibility: specimen.visibility, source: 'specimen-overall' };
  }
  const def = ownerDefault('images', owner);
  if (def != null) {
    return { visibility: def, source: 'user-default' };
  }
  return { visibility: SystemDefault, source: 'system-default' };
}

function specimenScalarOverride(
  field: Field,
  specimen: SpecimenLike,
): Visibility | null | undefined {
  switch (field) {
    case 'price':
      return specimen.visibility_price;
    case 'acquired_from':
      return specimen.visibility_acquired_from;
    case 'images':
      // Programmer error guard — matches the Go helper's
      // defensive fall-through.
      return null;
  }
}

function ownerDefault(field: Field, owner: OwnerLike): Visibility | null | undefined {
  const fd = owner.field_defaults;
  if (fd == null) {
    return null;
  }
  switch (field) {
    case 'price':
      return fd.price;
    case 'acquired_from':
      return fd.acquired_from;
    case 'images':
      return fd.images;
  }
}

File: internal/visibility/visibility.go
// Package visibility implements the per-field visibility resolution
// chain described in the V2 per-field visibility EPIC (mi-fo8). Given
// an already-loaded specimen, its owning user, and (for images) the
// photo, it picks the applicable [domain.Visibility] value plus the
// chain layer the value came from.
//
// The package is pure: no DB calls, no HTTP, no request context. The
// "is the viewer allowed to see this Visibility" check is a separate
// concern (see internal/authz and internal/api). Handlers call into
// this package to discover which value applies, then run the existing
// viewer-allows-visibility predicate against the result before deciding
// whether to redact a field from the response.
//
// Both [ResolveScalar] and [ResolveImage] stop at the first non-nil
// layer of their chain. No merging, no most-restrictive-of, no
// intersection — see mi-fo8 for the canonical description.
package visibility

import (
	"github.com/dickeyfPersonalProjects/minerals/internal/domain"
)

// Field discriminates the redactable specimen fields covered by the
// V2 launch (mi-fo8): two scalars and the images collection. The
// String form matches the keys used in users.field_defaults JSON.
type Field int

const (
	// FieldPrice selects the resolution chain for specimens.price_cents.
	FieldPrice Field = iota
	// FieldAcquiredFrom selects the chain for specimens.acquired_from.
	FieldAcquiredFrom
	// FieldImages selects the chain for the photos collection. Pass
	// this to [ResolveImage], not [ResolveScalar] — see ResolveScalar.
	FieldImages
)

// String returns the canonical lowercase name of the Field, matching
// the keys used in the users.field_defaults JSON column.
func (f Field) String() string {
	switch f {
	case FieldPrice:
		return "price"
	case FieldAcquiredFrom:
		return "acquired_from"
	case FieldImages:
		return "images"
	}
	return ""
}

// Source labels the layer of the resolution chain that produced a
// [Resolution]. Handlers may surface this in debug responses or
// log fields to explain why a field was redacted; nothing in the
// authorization path branches on it.
type Source string

const (
	// SourceImage means the photo's own Visibility override applied.
	// Only reachable from [ResolveImage].
	SourceImage Source = "image"
	// SourceSpecimenField means a specimen-level per-field override
	// applied — VisibilityPrice / VisibilityAcquiredFrom for the
	// scalar chains, VisibilityImages for the image chain.
	SourceSpecimenField Source = "specimen-field"
	// SourceSpecimenOverall means the specimen's overall Visibility
	// column applied. Only reachable from [ResolveImage]; the scalar
	// chain does not consult specimens.visibility (per mi-fo8).
	SourceSpecimenOverall Source = "specimen-overall"
	// SourceUserDefault means the owner's per-field default applied
	// (users.field_defaults).
	SourceUserDefault Source = "user-default"
	// SourceSystemDefault means no layer set a value; the conservative
	// system default applied. See [SystemDefault].
	SourceSystemDefault Source = "system-default"
)

// SystemDefault is the conservative fallback applied when no layer of
// the resolution chain set a value. Per mi-fo8, owner-only is the
// most-restrictive value and the V1→V2 default that keeps migrated
// rows from leaking on the wire post-V2; that maps onto the existing
// [domain.VisibilityPrivate] enum value.
const SystemDefault = domain.VisibilityPrivate

// Resolution is the output of the visibility chain: the value that
// applies and the layer it came from.
type Resolution struct {
	Visibility domain.Visibility
	Source     Source
}

// ResolveScalar resolves the visibility of a scalar redactable field
// — [FieldPrice] or [FieldAcquiredFrom] — by walking the chain
// documented in mi-fo8:
//
//  1. the specimen's per-field override (VisibilityPrice / VisibilityAcquiredFrom)
//  2. the owner's per-field default (users.field_defaults)
//  3. [SystemDefault]
//
// owner is the specimen's author; per mi-fo8 the per-user defaults map
// lives on the owning user, not the viewer.
//
// Passing [FieldImages] is a programmer error — image resolution
// requires the photo-level override and the specimen overall column
// that ResolveScalar does not touch. The call is handled defensively
// by skipping the specimen-field layer (there is no scalar override
// for images) and falling through to the owner default and system
// default, but callers SHOULD use [ResolveImage] for image fields.
func ResolveScalar(field Field, spec domain.Specimen, owner domain.User) Resolution {
	if v := specimenScalarOverride(field, spec); v != nil {
		return Resolution{Visibility: *v, Source: SourceSpecimenField}
	}
	if v := ownerDefault(field, owner); v != nil {
		return Resolution{Visibility: *v, Source: SourceUserDefault}
	}
	return Resolution{Visibility: SystemDefault, Source: SourceSystemDefault}
}

// ResolveImage resolves the visibility of a single photo by walking
// the image-specific chain documented in mi-fo8:
//
//  1. the photo's own Visibility override
//  2. the specimen's VisibilityImages override
//  3. the specimen's overall Visibility column
//  4. the owner's "images" default
//  5. [SystemDefault]
//
// The chain stops at the first non-nil layer. Per the mi-fo8 open
// question called out in the bead, an image override that LOOSENS
// against the specimen overall (e.g. public photo on a private
// specimen) is allowed by design — the helper implements the chain
// as written. If that policy reverses later, change happens here.
func ResolveImage(spec domain.Specimen, owner domain.User, img domain.Photo) Resolution {
	if img.Visibility != nil {
		return Resolution{Visibility: *img.Visibility, Source: SourceImage}
	}
	if spec.VisibilityImages != nil {
		return Resolution{Visibility: *spec.VisibilityImages, Source: SourceSpecimenField}
	}
	if spec.Visibility != "" {
		return Resolution{Visibility: spec.Visibility, Source: SourceSpecimenOverall}
	}
	if v := ownerDefault(FieldImages, owner); v != nil {
		return Resolution{Visibility: *v, Source: SourceUserDefault}
	}
	return Resolution{Visibility: SystemDefault, Source: SourceSystemDefault}
}

func specimenScalarOverride(field Field, spec domain.Specimen) *domain.Visibility {
	switch field {
	case FieldPrice:
		return spec.VisibilityPrice
	case FieldAcquiredFrom:
		return spec.VisibilityAcquiredFrom
	}
	return nil
}

func ownerDefault(field Field, owner domain.User) *domain.Visibility {
	fd := owner.FieldDefaults
	if fd == nil {
		return nil
	}
	switch field {
	case FieldPrice:
		return fd.Price
	case FieldAcquiredFrom:
		return fd.AcquiredFrom
	case FieldImages:
		return fd.Images
	}
	return nil
}
